src/main/scala/mesosphere/marathon/state/Container.scala

```diff
@@ -108,54 +84,150 @@ object Container {
       }
 
-      def networkHostPortValidator(d: Docker): Validator[PortMapping] =
+      def networkHostPortValidator(docker: Docker): Validator[PortMapping] =
         isTrue[PortMapping]("hostPort is required for BRIDGE mode.") { pm =>
-          d.network match {
-            case Some(Mesos.ContainerInfo.DockerInfo.Network.BRIDGE) => pm.hostPort.isDefined
+          docker.network match {
+            case Some(ContainerInfo.DockerInfo.Network.BRIDGE) => pm.hostPort.isDefined
             case _ => true
           }
         }
-    }
 
-    object PortMappings {
-      val portMappingsValidator: Validator[Seq[PortMapping]] = validator[Seq[PortMapping]] { portMappings =>
+      val portMappingsValidator = validator[Seq[PortMapping]] { portMappings =>
         portMappings is every(valid)
         portMappings is elementsAreUniqueByOptional(_.name, "Port names must be unique.")
       }
 
-      def validForDocker(d: Docker): Validator[Seq[PortMapping]] = validator[Seq[PortMapping]] { pm =>
-        pm is every(valid (PortMapping.networkHostPortValidator(d)))
+      def validForDocker(docker: Docker): Validator[Seq[PortMapping]] = validator[Seq[PortMapping]] { pm =>
+        pm is every(valid(PortMapping.networkHostPortValidator(docker)))
       }
     }
 
-    implicit val dockerValidator = validator[Docker] { docker =>
-      docker.image is notEmpty
-      docker.portMappings is optional(PortMappings.portMappingsValidator and PortMappings.validForDocker(docker))
+    val validDockerContainer: Validator[Container.Docker] = validator[Container.Docker] { docker =>
+      docker.portMappings is optional(PortMapping.portMappingsValidator and PortMapping.validForDocker(docker))
     }
   }
 
-  // We need validation based on the container type, but don't have dedicated classes. Therefore this approach manually
-  // delegates validation to the matching validator
   implicit val validContainer: Validator[Container] = {
     val validGeneralContainer = validator[Container] { container =>
       container.volumes is every(valid)
     }
 
-    val validDockerContainer: Validator[Container] = validator[Container] { container =>
-      container.docker is notEmpty
-      container.docker.each is valid
+    new Validator[Container] {
+      override def apply(container: Container): Result = container match {
+        case _: Mesos => Success
+        case docker: Docker => validate(docker)(Docker.validDockerContainer)
+      }
+    } and validGeneralContainer
+  }
+
+  /**
+    * An intermediate structure that parallels the external JSON API for containers.
+    * This allows for validation of all JSON input combinations,
+    * before they have been reduced to a single outcome of type Container
+    * which would cover up too many possible error conditions.
+    * Furthrmore, we can provide trivial Formats implementation via apply/unapply.
+    */
+  case class Mask(
```

Comment:
Did you try a solution similar to the example provided [here](http://stackoverflow.com/questions/17021847/noise-free-json-format-for-sealed-traits-with-play-2-2-library) and come up with a solution that doesn't require conversion? What's the exact problem you were facing?